### PR TITLE
D&c regression 115

### LINKED ISSUE #115: D&C Regression : Appointments per person
### Parent Task(if present):

D&C

###

**Description**:

Exam question:
”We need to use future demographic data to determine the number of appointments per person in SNEE”

‌

**Stakeholder request:**
(Andrew)

‌

---

**Objective**:

- Determine the most accurate (from k-fold cross validation) method to determine best fit (RMSE + MAE).
  - Refit the model on ALL data once the model is shown to generalise well.
  - Ensure it is using Sklearn pipeline, and other best practices.
  - Once best parameters are found, disable the hyperparameter tuning step, so that the runtime is reasonable in github CD/CI pipeline.
  - Save as ‘demographic-month-sklearn.pkl’ using joblib
    - We will save as an artifact on github at a later date

---

**Key Data Sources**:

- Primary Care appointments
- Primary care patient lists.

---

**Methodology/Approach**:

- Use ridge or Lasso regression.
- Only decompose the demographic data, not month.
- One hot encode or cyclically encode the month data (ensure it is not ordinal, so that 1 is not <12.
- Do not use polynomials.
- Provide interesting (useful) plots for the output.

---

**Subtasks**:

- Review current code (by Andrew)
- Apply [model/method] for analysis
- Interpret results and create visualizations
- Ensure pipeline is packaged well and documented.
- Write brief report on findings Markdown doc

---

### **Outputs**:

- [https://github.com/SNEE-ICS/PrimaryCareDemandAndCapacity](https://github.com/SNEE-ICS/PrimaryCareDemandAndCapacity "smartCard-inline")
- Branch Name
- Pull request #

### PR DESCRIPTION
**Pull request by IbrahimKhan07**
Closes #115

### CODE REVIEW

**Comment by IbrahimKhan07** [comment]
> getting there, but we can't merge in it's current state
> 
> need to add the PCA params, as that isn't part of the grid search, use the custom PCA class and the hyperparam grid to get those
> 
> also need to remove all the redundant timeseries analysis code
> 
> so basically, put the PCA params into the grid search and tidy up the whole notebook

**Comment by IbrahimKhan07** [comment]
> > getting there, but we can't merge in it's current state
> > need to add the PCA params, as that isn't part of the grid search, use the custom PCA class and the hyperparam grid to get those
> > also need to remove all the redundant timeseries analysis code
> > so basically, put the PCA params into the grid search and tidy up the whole notebook

Added the PCA-specific hyperparameters, in the parameter grid for each model. 
Removed the timeseries analysis code

I've cleaned this notebook, moved imports into one cell etc. But it would be better if we can have separate trello ticket for tidying up all the noteboks in one.

**Comment by AJarman** [comment]
Please add 4 PCA components (currently goes 2,3,5???) and remove 5. 
Time series analysis code is still in this notebook

**Review by AJarman — CHANGES_REQUESTED** [review]
- Please remove the time series analysis in section 2/3 trend analysis and decomposition.
- Note that the name defined for the output model is in the ticket/issue/trello card, change that.
- remove the gradient boosting/random forest models
- Automatically select the model with the best (lowest) RMSE by sorting

**Comment by IbrahimKhan07** [comment]
- Decomposition removed (time series)
- File saved as given name on trello (name of file defined as constant)
- Removed Random Forest and Gradient boosting models
- Best model is being automatically selected, instead manually selecting LR model
- Write up for plots
- Please check section 6, to be removed or keep it?

**Review by AJarman — CHANGES_REQUESTED** [review]
Looks good
Can you move the redundant code (from me) into appointments_models.py please

**Comment by IbrahimKhan07** [comment]
Redundant code moved to specified file

**Review by AJarman — APPROVED** [review]
LGTM 🌮
